File: integrations/kagent/appa-kagent-adk-go/discovery.go
package appakagentadk

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"time"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

const discoveryMaxTools = 10_000
const discoveryMaxBytes = 10 * 1024 * 1024

var discoveredToolName = regexp.MustCompile(`^[A-Za-z0-9_.-]+$`)

type DiscoveryStatus string

const (
	DiscoveryComplete    DiscoveryStatus = "complete"
	DiscoveryPartial     DiscoveryStatus = "partial"
	DiscoveryUnavailable DiscoveryStatus = "unavailable"
)

// Discovery is metadata evidence, not permission to execute the returned tools.
type Discovery struct {
	Tools  []*mcp.Tool
	Status DiscoveryStatus
	Detail string
}

// Discover reuses the host's authenticated session. It never creates a transport
// or invokes a tool, and the host remains responsible for session cleanup.
func Discover(ctx context.Context, session *mcp.ClientSession, filter []string) (Discovery, error) {
	deadline, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()
	found := make(map[string]*mcp.Tool)
	selected := make(map[string]bool)
	for _, name := range filter {
		selected[name] = true
	}
	result := func(status DiscoveryStatus, detail string) Discovery {
		tools := make([]*mcp.Tool, 0, len(found))
		for name, tool := range found {
			if len(selected) == 0 || selected[name] {
				tools = append(tools, tool)
			}
		}
		sort.Slice(tools, func(i, j int) bool { return tools[i].Name < tools[j].Name })
		return Discovery{Tools: tools, Status: status, Detail: detail}
	}
	cursor := ""
	cursors := make(map[string]bool)
	size := 0
	for {
		page, err := session.ListTools(deadline, &mcp.ListToolsParams{Cursor: cursor})
		if err != nil {
			if ctx.Err() != nil {
				return Discovery{}, ctx.Err()
			}
			status := DiscoveryUnavailable
			if len(found) != 0 {
				status = DiscoveryPartial
			}
			// Transport errors can quote credentials; retain no exception text.
			return result(status, "MCP metadata discovery did not complete"), nil
		}
		if page == nil {
			return Discovery{}, fmt.Errorf("MCP tool listing has no result")
		}
		encoded, err := json.Marshal(page)
		if err != nil {
			return Discovery{}, fmt.Errorf("MCP tool listing contains invalid metadata")
		}
		size += len(encoded)
		if size > discoveryMaxBytes || len(found)+len(page.Tools) > discoveryMaxTools {
			return Discovery{}, fmt.Errorf("MCP tool listing exceeds the metadata discovery limit")
		}
		for _, tool := range page.Tools {
			if tool == nil || !discoveredToolName.MatchString(tool.Name) || found[tool.Name] != nil {
				return Discovery{}, fmt.Errorf("MCP tool listing contains an invalid or duplicate tool")
			}
			found[tool.Name] = tool
		}
		cursor = page.NextCursor
		if cursor == "" {
			break
		}
		if cursors[cursor] {
			return Discovery{}, fmt.Errorf("MCP tool listing repeats a pagination cursor")
		}
		cursors[cursor] = true
	}
	for name := range selected {
		if found[name] == nil {
			return Discovery{}, fmt.Errorf("MCP tool filter names tools absent from the complete listing")
		}
	}
	return result(DiscoveryComplete, ""), nil
}

File: integrations/kagent/appa-kagent-adk/src/appa_kagent_adk/discovery.py
"""Bounded metadata discovery using the host's already-authenticated MCP session.

This module never opens a connection, invokes a tool, or grants policy coverage.
The host retains ownership of transports, credentials, and session cleanup.
"""

from __future__ import annotations

import asyncio
import re
from dataclasses import dataclass
from typing import Literal

from mcp import ClientSession
from mcp.types import Tool

from .config_guard import ConfigRefused

MAX_TOOLS = 10_000
MAX_BYTES = 10 * 1024 * 1024
TIMEOUT_SECONDS = 30


@dataclass(frozen=True)
class Discovery:
    tools: tuple[Tool, ...]
    status: Literal["complete", "partial", "unavailable"]
    detail: str | None = None


async def discover_toolset(toolset, readonly_context=None) -> Discovery:
    """Probe a stock toolset through its authenticated session boundary.

    The two pinned ADK lanes expose this same session operation. It retains
    kagent's TLS configuration, static headers, token propagation and OAuth.
    The outer deadline also bounds connection setup, not only list_tools.
    Validation errors return through the callback as data because ADK otherwise
    wraps them in ConnectionError, which would misreport known errors as unknown.
    """
    selected = toolset.tool_filter
    if selected is not None and not isinstance(selected, list):
        raise ConfigRefused("MCP discovery requires a name-list filter or no filter")

    async def inspect(session):
        try:
            return await discover(session, selected)
        except ConfigRefused as error:
            return error

    try:
        result = await asyncio.wait_for(
            toolset._execute_with_session(inspect, "MCP metadata discovery did not complete", readonly_context),
            timeout=TIMEOUT_SECONDS,
        )
    except Exception:
        return Discovery((), "unavailable", "MCP connection or metadata discovery did not complete")
    if isinstance(result, ConfigRefused):
        raise result
    return result


async def discover(session: ClientSession, tool_filter: list[str] | None = None) -> Discovery:
    """Enumerate every page before applying the host's optional name filter.

    A failed listing is incomplete evidence, not an empty complete inventory.
    Invalid listings raise ConfigRefused; exception messages from a transport are
    never copied into diagnostics because they can contain authentication data.
    """
    found: dict[str, Tool] = {}
    selected = set(tool_filter) if tool_filter else None
    size = 0

    async def pages() -> None:
        nonlocal size
        cursor: str | None = None
        cursors: set[str] = set()
        while True:
            page = await session.list_tools(cursor=cursor)
            size += len(page.model_dump_json().encode("utf-8"))
            if size > MAX_BYTES or len(found) + len(page.tools) > MAX_TOOLS:
                raise ConfigRefused("MCP tool listing exceeds the metadata discovery limit")
            for tool in page.tools:
                if re.fullmatch(r"[A-Za-z0-9_.-]+", tool.name) is None:
                    raise ConfigRefused("MCP tool listing contains an invalid tool name")
                if tool.name in found:
                    raise ConfigRefused("MCP tool listing contains a duplicate tool name")
                found[tool.name] = tool
            cursor = page.nextCursor
            if not cursor:
                return
            if cursor in cursors:
                raise ConfigRefused("MCP tool listing repeats a pagination cursor")
            cursors.add(cursor)

    def tools() -> tuple[Tool, ...]:
        return tuple(found[name] for name in sorted(found) if selected is None or name in selected)

    try:
        await asyncio.wait_for(pages(), timeout=TIMEOUT_SECONDS)
    except ConfigRefused:
        raise
    except Exception:
        # Cancellation deliberately propagates: it is not a connection failure.
        return Discovery(tools(), "partial" if found else "unavailable", "MCP metadata discovery did not complete")
    if selected is not None and selected.difference(found):
        raise ConfigRefused("MCP tool filter names tools absent from the complete listing")
    return Discovery(tools(), "complete")
